# Processing Vancouver city Census 2016 dataset

## Reading data Input

Using read_csv (function of pandas) to read the input file. The **encoding='ISO-8859-1** option permits read the input file properly. The **index_col=0** option, tells to read_csv function that the first column in the data set is the index column. 

In [185]:
import numpy as np
import pandas as pd

df = pd.read_csv('data/CensusLocalAreaProfiles2016.csv', encoding='ISO-8859-1', index_col=0)


---

## Determining existence of data in data frame(input file)

In [186]:
df.empty

False

Due the function **empty** (applied in the data frame created) is False, The dataframe has data. This step is important in order to check if the data was copied correctly in the hard disk, transference issues or there are an inconvinient in generation of data frame.

---

## Finding dimensions of the dataframe


In [187]:
df.shape

(5589, 25)

According to the function **shape** applied to the data frame, the data has 5589 rows and 26 columns

---

## Size of the Input File  (CensusLocalAreaProfiles2016.csv)

In [188]:
!dir data\CensusLocalAreaProfiles2016.csv

 Volume in drive C is Windows8_OS
 Volume Serial Number is 4623-A251

 Directory of C:\Users\Jhonny\Google Drive\Coding\git\Data Analysis Project-Population vs Schools\data

2019-11-14  08:01 PM           628,241 CensusLocalAreaProfiles2016.csv
               1 File(s)        628,241 bytes
               0 Dir(s)  324,695,138,304 bytes free


Determining the size of the input file, permits check posible data lost in data copying.

---

## Checking permissions in the file

!icacls data\CensusLocalAreaProfiles2016.csv

Checking permissions in the file avoid possible errors when creating data or running tasks.

(I) - permission inherited from parent container
 F - full access

---

## Visualization of the dataframe

In [189]:
df.head()

Variable Arbutus-Ridge   \
ID                                                                     
1    Total - Age groups and average age of the pop...         15,295   
2                                       0 to 14 years           2015   
3                                        0 to 4 years            455   
4                                        5 to 9 years            685   
5                                      10 to 14 years            880   

   Downtown  Dunbar-Southlands  Fairview  Grandview-Woodland   \
ID                                                              
1     62,030             21,425    33,620              29,175   
2       4000               3545      2580                3210   
3       2080                675      1240                1320   
4       1105               1225       760                1025   
5        810               1650       580                 865   

   Hastings-Sunrise  Kensington-Cedar Cottage  Kerrisdale  Killarney   ...  \
ID                                                                     ...   
1             34,575                    49,325      13,975     29,325  ...   
2               4595                      7060        1880       4185  ...   
3               1510                      2515         430       1300  ...   
4               1560                      2390         600       1400  ...   
5               1525                      2160         845       1485  ...   

   Riley Park  Shaughnessy  South Cambie  Strathcona  Sunset   \
ID                                                              
1       22,555        8,430         7,970      12,585  36,500   
2         3415         1175          1105        1065    5460   
3         1175          270           360         360    1695   
4         1160          405           365         365    1780   
5         1080          500           375         340    1985   

   Victoria-Fraserview West End  West Point Grey  Vancouver CSD   \
ID                                                                 
1               31,065    47,200           13,065        631,485   
2                 3790      1945             1900          70530   
3                 1175       965              420          24615   
4                 1210       560              670          22905   
5                 1410       415              810          23010   

   Vancouver CMA   
ID                 
1       2,463,430  
2          362110  
3          116270  
4          122730  
5          123110  

[5 rows x 25 columns]

The function **head()** gives an idea that how the data looks like.

---

## Checking columns in dataframe

Using the function **columns**, the name of each column can be checked.

In [190]:
df.columns

Index(['Variable', 'Arbutus-Ridge ', 'Downtown ', 'Dunbar-Southlands ',
       'Fairview ', 'Grandview-Woodland ', 'Hastings-Sunrise ',
       'Kensington-Cedar Cottage ', 'Kerrisdale ', 'Killarney ', 'Kitsilano ',
       'Marpole ', 'Mount Pleasant ', 'Oakridge ', 'Renfrew-Collingwood ',
       'Riley Park ', 'Shaughnessy ', 'South Cambie ', 'Strathcona ',
       'Sunset ', 'Victoria-Fraserview', 'West End ', 'West Point Grey ',
       'Vancouver CSD ', 'Vancouver CMA '],
      dtype='object')

 A space in the end of aech column's name was found. This space should be deleted.

---

### Deleting unnecessary spaces on columns names

Using **rename** function, the spaces in the end of each name of the column can be deleted by renaming it

In [191]:
df.rename(
    columns={
        'Arbutus-Ridge ' : 'Arbutus-Ridge', 
        'Downtown ' : 'Downtown', 
        'Dunbar-Southlands ' : 'Dunbar-Southlands',
        'Fairview ' : 'Fairview', 
        'Grandview-Woodland ' : 'Grandview-Woodland',
        'Hastings-Sunrise ' : 'Hastings-Sunrise',
        'Kensington-Cedar Cottage ' : 'Kensington-Cedar Cottage' ,
        'Kerrisdale ' : 'Kerrisdale', 
        'Killarney ' : 'Killarney', 
        'Kitsilano ' : 'Kitsilano',
        'Marpole ' : 'Marpole',
        'Mount Pleasant ' : 'Mount Pleasant',
        'Oakridge ' : 'Oakridge',
        'Renfrew-Collingwood ' : 'Renfrew-Collingwood',
        'Riley Park ' : 'Riley Park', 
        'Shaughnessy ' : 'Shaughnessy', 
        'South Cambie ' : 'South Cambie' , 
        'Strathcona ' : 'Strathcona' ,
        'Sunset ' : 'Sunset' ,  
        'West End ' : 'West End' , 
        'West Point Grey ' : 'West Point Grey',
        'Vancouver CSD ' : 'Vancouver CSD' , 
        'Vancouver CMA ' : 'Vancouver CMA' 
    }, inplace=True
)

### Quality control of the edition

The function **Columns** is called again after the edition of spaces in order to check that the names was written with out errors.

In [192]:
df.columns

Index(['Variable', 'Arbutus-Ridge', 'Downtown', 'Dunbar-Southlands',
       'Fairview', 'Grandview-Woodland', 'Hastings-Sunrise',
       'Kensington-Cedar Cottage', 'Kerrisdale', 'Killarney', 'Kitsilano',
       'Marpole', 'Mount Pleasant', 'Oakridge', 'Renfrew-Collingwood',
       'Riley Park', 'Shaughnessy', 'South Cambie', 'Strathcona', 'Sunset',
       'Victoria-Fraserview', 'West End', 'West Point Grey', 'Vancouver CSD',
       'Vancouver CMA'],
      dtype='object')

The spaces after the name of each columns was deleted. Deleting spaces is important due it can avoid possible future errors in the data processing.

---

## Information of data types in data frame

The function **dtypes** permits see data types of each column of the data frame. Knowing data types in data frame could avoid possible errors or unexpected results in the future.

In [193]:
df.dtypes


Variable                    object
Arbutus-Ridge               object
Downtown                    object
Dunbar-Southlands           object
Fairview                    object
Grandview-Woodland          object
Hastings-Sunrise            object
Kensington-Cedar Cottage    object
Kerrisdale                  object
Killarney                   object
Kitsilano                   object
Marpole                     object
Mount Pleasant              object
Oakridge                    object
Renfrew-Collingwood         object
Riley Park                  object
Shaughnessy                 object
South Cambie                object
Strathcona                  object
Sunset                      object
Victoria-Fraserview         object
West End                    object
West Point Grey             object
Vancouver CSD               object
Vancouver CMA               object
dtype: object

All columns are type **object** (used for numeric, non-numeric and text values).

---

## Finding nulls

The function **info()** shows a complete information of the data frame, including non-null values.

In [194]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5589 entries, 1 to 5493
Data columns (total 25 columns):
Variable                    5493 non-null object
Arbutus-Ridge               5493 non-null object
Downtown                    5493 non-null object
Dunbar-Southlands           5493 non-null object
Fairview                    5493 non-null object
Grandview-Woodland          5493 non-null object
Hastings-Sunrise            5493 non-null object
Kensington-Cedar Cottage    5493 non-null object
Kerrisdale                  5493 non-null object
Killarney                   5493 non-null object
Kitsilano                   5493 non-null object
Marpole                     5493 non-null object
Mount Pleasant              5493 non-null object
Oakridge                    5493 non-null object
Renfrew-Collingwood         5493 non-null object
Riley Park                  5493 non-null object
Shaughnessy                 5493 non-null object
South Cambie                5493 non-null object
Strathcona      

As reported by the function **info()** there are 5589 entries (can be describe as 5589 rows) and 26 columns. It is the same dimension of the data frame reported by the function **dimensions** in the step three of this workfllow. A simple subtration of 5494 to 5589, shows that there are 95 null objects in that column. Similar operation in the rest of the columns, shows 96 null objects in each column. For now, no any actions it will be taken with the null objects.

---

## Describing all data types

Although the function **describe** is more useful in numeric values and not in objects, the application of this object can give some infomration in this step.

In [195]:
df.describe(include='all')

Variable Arbutus-Ridge Downtown Dunbar-Southlands  \
count                     5493          5493     5493              5493   
unique                    2278           756     1222               847   
top         English and French             0        0                 0   
freq                        11          2742     1910              2607   

       Fairview Grandview-Woodland Hastings-Sunrise Kensington-Cedar Cottage  \
count      5493               5493             5493                     5493   
unique     1014                953             1005                     1100   
top           0                  0                0                        0   
freq       2284               2261             2327                     2190   

       Kerrisdale Killarney  ... Riley Park Shaughnessy South Cambie  \
count        5493      5493  ...       5493        5493         5493   
unique        739       928  ...        857         615          589   
top             0         0  ...          0           0            0   
freq         2798      2317  ...       2506        2961         2893   

       Strathcona Sunset Victoria-Fraserview West End West Point Grey  \
count        5493   5493                5493     5493            5493   
unique        678   1008                 964     1158             719   
top             0      0                   0        0               0   
freq         2676   2260                2448     2032            2726   

       Vancouver CSD Vancouver CMA  
count           5493          5493  
unique          2099          2553  
top                0             0  
freq            1187           889  

[4 rows x 25 columns]

In the stats returned by the function **count** shows the total number of non-null objects. In adition, ID's unique numbers are checked too.

---

## Continuing Finding nulls

Mising values in a data frame can be detected using the function **isna()** . in the ID column of the data frame **isna()** function was applied and it was saved in a variable called **_containsNulls_**.

With the **shape()** function, nulls values in that colum can be counted. 

As the result obtain with the values using **info()** function in a step above, the function **Shape()** shows a total of 95 null objects.

It can be apply in any other column of the data frame

In [196]:
containsNulls1 = df[df.Downtown.isna()] 

In [197]:
containsNulls1.shape[0]

96

In this part of the process, it is not necessary ubicate in the data frame the value of the null objects.

---

## Writing data (output)

Finally, using **to_csv** function the dataset checked and cleaned is written in format .csv

In [198]:
df.to_csv('data/CensusLocalAreaProfiles2016_cleaned.csv')